walnuts_cpp/include/walnuts/adaptive_walnuts.hpp

```diff
@@ -0,0 +1,399 @@
+    mass_cfg_(mass_cfg), 
+    step_cfg_(step_cfg),
+    walnuts_cfg_(walnuts_cfg),
+    rand_(rng),
+    logp_grad_(logp_grad),
+    theta_(theta_init),
+    iteration_(0),
+    step_adapt_handler_(step_cfg.step_size_init_, step_cfg.accept_rate_target_,
+			step_cfg.iteration_offset_, step_cfg.learning_rate_,
+			step_cfg.decay_rate_),
+    mass_adapt_var_(0.0, theta_init.size()),  // init overridden later
+    mass_adapt_prec_(0.0, theta_init.size())
+  {
+    S smoothing = 0.05;
+    Vec<S> zero = Vec<S>::Zero(theta_init.size());
+    Vec<S> smooth_vec = Vec<S>::Constant(theta_init.size(), smoothing);
+    Vec<S> sqrt_abs_grad_init = grad(logp_grad, theta_init).array().abs().sqrt();
+    Vec<S> init_prec = (1 - smoothing) * sqrt_abs_grad_init + smooth_vec;
+    Vec<S> init_var = init_prec.array().inverse().matrix();
+    mass_adapt_var_ = OnlineMoments<S>(0.98, mass_cfg.iteration_offset_,
+				       zero, init_var);
+    mass_adapt_prec_ = OnlineMoments<S>(0.98, mass_cfg.iteration_offset_,
+					zero, init_prec);
+  }
+
+
+  /**
+   * @brief Return the next state from warmup.  
+   *
+   * This method should be called a number of time equal to the
+   * number of warmup iterations desired.  These warmup draws are
+   * *not* drawn from a Markov chain and are not valid for inference.
+   * After warmup, call `sampler()` to return a sampler that fixes the
+   * tuning parameters and provides a proper Markov chain.
+   * 
+   * @return The next warmup state.
+   */
+  const Vec<S> operator()() {
+    Vec<S> inv_mass_est_var = mass_adapt_var_.variance().array();
+    Vec<S> inv_mass_est_prec = mass_adapt_prec_.variance().array().inverse()
+      .matrix();
+    Vec<S> inv_mass = (inv_mass_est_var.array() * inv_mass_est_prec.array())
+      .sqrt().matrix();
+    Vec<S> mass = inv_mass.array().inverse().matrix();
+    Vec<S> chol_mass = mass.array().sqrt().matrix();
+    Vec<S> grad_select;
+    theta_ = transition_w(rand_, logp_grad_, inv_mass, chol_mass,
+        		  step_adapt_handler_.step_size(),
+                          walnuts_cfg_.max_nuts_depth_,
+                          std::move(theta_), grad_select, walnuts_cfg_.log_max_error_,
+                          step_adapt_handler_);
+    double discount_factor = 1.0 - 1.0 / (mass_cfg_.iteration_offset_ + iteration_);
+    mass_adapt_var_.set_discount_factor(discount_factor);
+    mass_adapt_var_.observe(theta_);
+    mass_adapt_prec_.set_discount_factor(discount_factor);
+    mass_adapt_prec_.observe(grad_select); 
+    ++iteration_;
+    return theta_;
+  }
+
+  /**
+   * @brief Return a WALNUTS sampler with the current tuning parameter
+   * estimates. 
+   *
+   * The returned sampler forms a proper Markov chain.
+   *
+   * @return The WALNUTS sampler with current tuning parameter estimates. 
+   */
+  WalnutsSampler<F, S, RNG> sampler() {
+    return WalnutsSampler<F, S, RNG>(
+        rand_,
+        logp_grad_,
+        theta_,
+        mass_adapt_var_.variance(),
```

Comment:
Shouldn't the mass matrix for the new sampler also use `mass_adapt_prec_`?

Reply:
Yes, it should.  I have no idea how I let that slip through!

Reply:
I pushed a fix that encapsulates the mass matrix estimator to avoid cutting-and-pasting and then added the extra config to the mass estimator config. The sampling code is much cleaner and the mass matrix estimators are better.